### ISSUE TITLE
MySQL SET @variable from SELECT, must not have SET keyword after conversion

### ISSUE THREAD

**Issue by krasandr** [darold/ora2pg#605]
There is a problem with conversion of MySQL @variables, when they are assigned by result of SELECT. For example original MySQL function code:
```
| f_artstatus | STRICT_TRANS_TABLES,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION | CREATE DEFINER=`root`@`%` FUNCTION `f_artstatus`(p_IvfArtID int,
IsCancel int) RETURNS int(11)
BEGIN
  SET @IvfArtID := p_IvfArtID;
  SET @IsCancel := IsCancel;
  SELECT
    a.IvfArtFormID,
    a.IvfArtStatusID,
    arts.IsCanceled INTO @IvfFormID, @OldArtStatusID, @IsOldCanceled
  FROM ivf_arts a,
       ivf_const_art_statuses arts
  WHERE a.IvfArtID = @IvfArtID
  AND a.IvfArtStatusID = arts.IvfArtStatusID;

  IF (@OldArtStatusID = 16)
    THEN
    RETURN 16;
  END IF;

  SET @PriklichilaDonorka :=
  IFNULL((SELECT
      1 AS b
    FROM ivf_stimulations s
      INNER JOIN ivf_punctions p
        ON s.IvfStimID = p.IvfStimID
    WHERE s.IvfArtID = @IvfArtID
    AND NOT EXISTS (SELECT
        1
      FROM v_ocells o
      WHERE o.IvfArtID = s.IvfArtID
      AND o.IvfOStatusID = 1)), 0);

  IF (@PriklichilaDonorka = 1
    AND @IvfFormID = 4)
    THEN
    RETURN 15;
  END IF;
  
  SET @IsFertiliz := IFNULL((SELECT
      IF(COUNT(*) = 0, 0, 1)
    FROM ivf_stimulations s,
         ivf_ins_fertilizations c,
         ivf_ins_fertilizations_ivf_s_cells fc
    WHERE s.IvfStimID = c.IvfStimID
    AND fc.IvfInsFerID = c.IvfInsFerID
    AND s.IvfArtID = @IvfArtID), 0);
  IF (@IsFertiliz = 1)
    THEN
    RETURN 13;
  END IF;

  SET @IsEmbrioDev := IFNULL((SELECT
      IF(COUNT(*) > 0, 1, 0)
    FROM v_ocells o
    WHERE o.IvfArtID = @IvfArtID
    AND o.IvfOStatusID = 1), 0);

  IF (@IsEmbrioDev = 1)
    THEN
    RETURN 7;
  END IF;
  
  SET @IsTransfer := (SELECT
      1
    FROM v_ocells v
    WHERE v.IvfArtID = @IvfArtID
    AND v.IvfTransferID IS NOT NULL
    LIMIT 1);

  
  SET @IsFRZEmbrions := (SELECT
      1
    FROM v_ocells v
    WHERE v.IvfArtID = @IvfArtID
    AND v.IvfCryoEggID IS NOT NULL
    LIMIT 1);

  
  IF (@IsTransfer = 1
    AND @IsFRZEmbrions = 1)
    THEN
    RETURN 12;
  END IF;

  IF (@IsFRZEmbrions = 1)
    THEN
    RETURN 11;
  END IF;

  IF (@IsTransfer = 1)
    THEN
    RETURN 10;
  END IF;

  IF (@IsCancel = 1
    AND @IvfFormID = 3)
    THEN
    RETURN 9;
  END IF;
  
  SET @IsCancelAfterPunc :=
  IFNULL((SELECT
      1 AS b
    FROM ivf_stimulations s,
         ivf_punctions p
    WHERE s.IvfStimID = p.IvfStimID
    AND s.IvfArtID = @IvfArtID
    AND NOT EXISTS (SELECT
        1
      FROM v_ocells o
      WHERE o.IvfArtID = s.IvfArtID
      AND o.IvfOStatusID != 4)), 0);

  IF (@IsCancelAfterPunc = 1)
    THEN
    RETURN 8;
  END IF;

  IF (@IsOldCanceled = 1)
    THEN
    SET @IsCancel = IF(@IsCancel = 0, 1, 0);
  END IF;

  IF (@IsCancel = 1
    AND @IvfFormID = 5)
    THEN
    RETURN 5;
  END IF;

  IF (@IsCancel = 1
    AND @IvfFormID = 3)
    THEN
    RETURN 9;
  END IF;

  IF (@IsCancel = 1)
    THEN
    RETURN 6;
  END IF;

  SET @IsEmbrCancel := (SELECT
  
      IF(COUNT(*) > 0
      AND COUNT(*) = SUM(IF(iod.IvfOStatusID = 4, 1, 0)),
      1, 0) AS Result

    FROM ivf_embryo_protocols iep
      INNER JOIN ivf_ocells_developments iod
        ON iep.EmbryoProtocolID = iod.EmbryoProtocolID

    WHERE iep.IvfArtID = @IvfArtID);

  IF (@IsEmbrCancel = 1) THEN
    RETURN 9;
  END IF;
  
  SET @IsStim := IFNULL((SELECT
      1
    FROM ivf_stimulations s
    WHERE s.IvfArtID = @IvfArtID), 0);

  IF (@IsStim = 1
    AND @IvfFormID = 5)
    THEN
    RETURN 4;
  END IF;

  IF (@IsStim = 1)
    THEN
    RETURN 3;
  END IF;

  IF (@IvfFormID = 2)
    THEN
    SET @IsIschakvane := (SELECT
        1
      FROM v_ocells o
      WHERE o.IvfArtID = @IvfArtID
      LIMIT 1);

    IF (@IsIschakvane != 1)
      THEN
      RETURN 2;
    END IF;
  END IF;

  RETURN 1;

END | utf8                 | utf8_general_ci      | latin1_swedish_ci  |
```

Converted with Ora2Pg function code is:
```
CREATE OR REPLACE FUNCTION f_artstatus (p_IvfArtID integer, IsCancel integer) RETURNS integer AS $body$
DECLARE
IvfArtID integer;
IsCancel integer;
PriklichilaDonorka integer;
IsFertiliz integer;
IsEmbrioDev integer;
IsTransfer integer;
IsFRZEmbrions integer;
IsEmbrCancel integer;
IsStim integer;
IsIschakvane integer;
IvfFormID varchar;
OldArtStatusID varchar;
IsOldCanceled varchar;

BEGIN

  IvfArtID := p_IvfArtID;
  IsCancel := IsCancel;

  SELECT
    a.IvfArtFormID,
    a.IvfArtStatusID,
    arts.IsCanceled INTO IvfFormID, OldArtStatusID, IsOldCanceled
  FROM ivf_arts a,
       ivf_const_art_statuses arts
  WHERE a.IvfArtID = IvfArtID
  AND a.IvfArtStatusID = arts.IvfArtStatusID;
  
  IF (OldArtStatusID = 16)
    THEN
    RETURN 16;
  END IF;

  SET PriklichilaDonorka = COALESCE((SELECT
      1 AS b
    FROM ivf_stimulations s
      INNER JOIN ivf_punctions p
        ON s.IvfStimID = p.IvfStimID
    WHERE s.IvfArtID = IvfArtID
    AND NOT EXISTS (SELECT
        1
      FROM v_ocells o
      WHERE o.IvfArtID = s.IvfArtID
      AND o.IvfOStatusID = 1)), 0);

  IF (PriklichilaDonorka = 1
    AND IvfFormID = 4)
    THEN
    RETURN 15;
  END IF;

  
  SET IsFertiliz = COALESCE((SELECT
      CASE WHEN COUNT(*) = 0 THEN 0 ELSE  1 END 
    FROM ivf_stimulations s, ivf_ins_fertilizations c,
         ivf_ins_fertilizations_ivf_s_cells fc
    WHERE s.IvfStimID = c.IvfStimID
    AND fc.IvfInsFerID = c.IvfInsFerID
    AND s.IvfArtID = IvfArtID), 0);
  IF (IsFertiliz = 1)
    THEN
    RETURN 13;
  END IF;

  
  SET IsEmbrioDev = COALESCE((SELECT
      CASE WHEN COUNT(*) > 0 THEN 1 ELSE  0 END 
    FROM v_ocells o
    WHERE o.IvfArtID = IvfArtID
    AND o.IvfOStatusID = 1), 0);

  IF (IsEmbrioDev = 1)
    THEN
    RETURN 7;
  END IF;


  
  SET IsTransfer = (SELECT
      1
    FROM v_ocells v
    WHERE v.IvfArtID = IvfArtID
    AND v.IvfTransferID IS NOT NULL
    LIMIT 1);

  
  SET IsFRZEmbrions = (SELECT
      1
    FROM v_ocells v
    WHERE v.IvfArtID = IvfArtID
    AND v.IvfCryoEggID IS NOT NULL
    LIMIT 1);

  
  IF (IsTransfer = 1
    AND IsFRZEmbrions = 1)
    THEN
    RETURN 12;
  END IF;

  
  IF (IsFRZEmbrions = 1)
    THEN
    RETURN 11;
  END IF;


  
  IF (IsTransfer = 1)
    THEN
    RETURN 10;
  END IF;

  
  IF (IsCancel = 1
    AND IvfFormID = 3)
    THEN
    RETURN 9;
  END IF;



  
  SET IsCancelAfterPunc :=
  COALESCE((SELECT
      1 AS b
    FROM ivf_stimulations s, ivf_punctions p
    WHERE s.IvfStimID = p.IvfStimID
    AND s.IvfArtID = IvfArtID
    AND NOT EXISTS (SELECT
        1
      FROM v_ocells o
      WHERE o.IvfArtID = s.IvfArtID
      AND o.IvfOStatusID != 4)), 0);

  IF (IsCancelAfterPunc = 1)
    THEN
    RETURN 8;
  END IF;



  
  IF (IsOldCanceled = 1)
    THEN
    IsCancel := CASE WHEN IsCancel = 0 THEN 1 ELSE  0 END ;
  END IF;


  
  
  IF (IsCancel = 1
    AND IvfFormID = 5)
    THEN
    RETURN 5;
  END IF;


  
  IF (IsCancel = 1
    AND IvfFormID = 3)
    THEN
    RETURN 9;
  END IF;

  IF (IsCancel = 1)
    THEN
    RETURN 6;
  END IF;

  SET IsEmbrCancel = (SELECT
      
      
      CASE WHEN COUNT(*) > 0      AND COUNT(*) = SUM(CASE WHEN iod.IvfOStatusID = 4 THEN 1 ELSE  0 END ) THEN 1 ELSE  0 END  AS Result

    FROM ivf_embryo_protocols iep
      INNER JOIN ivf_ocells_developments iod
        ON iep.EmbryoProtocolID = iod.EmbryoProtocolID

    WHERE iep.IvfArtID = IvfArtID);

  IF (IsEmbrCancel = 1) THEN
    RETURN 9;
  END IF;

  
  SET IsStim = COALESCE((SELECT
      1
    FROM ivf_stimulations s
    WHERE s.IvfArtID = IvfArtID), 0);


  
  IF (IsStim = 1
    AND IvfFormID = 5)
    THEN
    RETURN 4;
  END IF;

  
  IF (IsStim = 1)
    THEN
    RETURN 3;
  END IF;

  
  IF (IvfFormID = 2)
    THEN
    SET IsIschakvane = (SELECT
        1
      FROM v_ocells o
      WHERE o.IvfArtID = IvfArtID
      LIMIT 1);

    IF (IsIschakvane != 1)
      THEN
      RETURN 2;
    END IF;
  END IF;


  
  RETURN 1;

END;
$body$
LANGUAGE PLPGSQL
SECURITY DEFINER
;
-- REVOKE ALL ON FUNCTION f_artstatus (p_IvfArtID integer, IsCancel integer) FROM PUBLIC;
```

If I simply remove every word "SET" before assignment, when @variable is assigned by result of SELECT, everything works (except that one variable IsCancelAfterPunc was not defined in DECLARE):
```
CREATE OR REPLACE FUNCTION f_artstatus (p_IvfArtID integer, IsCancel integer) RETURNS integer AS $body$
DECLARE
IvfArtID integer;
IsCancel integer;
PriklichilaDonorka integer;
IsFertiliz integer;
IsEmbrioDev integer;
IsTransfer integer;
IsFRZEmbrions integer;
IsEmbrCancel integer;
IsStim integer;
IsIschakvane integer;
IvfFormID varchar;
OldArtStatusID varchar;
IsOldCanceled varchar;
IsCancelAfterPunc varchar;

BEGIN

  IvfArtID := p_IvfArtID;
  IsCancel := IsCancel;
  
  SELECT
    a.IvfArtFormID,
    a.IvfArtStatusID,
    arts.IsCanceled INTO IvfFormID, OldArtStatusID, IsOldCanceled
  FROM ivf_arts a,
       ivf_const_art_statuses arts
  WHERE a.IvfArtID = IvfArtID
  AND a.IvfArtStatusID = arts.IvfArtStatusID;

  
  IF (OldArtStatusID = 16)
    THEN
    RETURN 16;
  END IF;


  
  PriklichilaDonorka = COALESCE((SELECT
      1 AS b
    FROM ivf_stimulations s
      INNER JOIN ivf_punctions p
        ON s.IvfStimID = p.IvfStimID
    WHERE s.IvfArtID = IvfArtID
    AND NOT EXISTS (SELECT
        1
      FROM v_ocells o
      WHERE o.IvfArtID = s.IvfArtID
      AND o.IvfOStatusID = 1)), 0);

  IF (PriklichilaDonorka = 1
    AND IvfFormID = 4)
    THEN
    RETURN 15;
  END IF;

  
  IsFertiliz = COALESCE((SELECT
      CASE WHEN COUNT(*) = 0 THEN 0 ELSE  1 END 
    FROM ivf_stimulations s, ivf_ins_fertilizations c,
         ivf_ins_fertilizations_ivf_s_cells fc
    WHERE s.IvfStimID = c.IvfStimID
    AND fc.IvfInsFerID = c.IvfInsFerID
    AND s.IvfArtID = IvfArtID), 0);
  IF (IsFertiliz = 1)
    THEN
    RETURN 13;
  END IF;

  
  IsEmbrioDev = COALESCE((SELECT
      CASE WHEN COUNT(*) > 0 THEN 1 ELSE  0 END 
    FROM v_ocells o
    WHERE o.IvfArtID = IvfArtID
    AND o.IvfOStatusID = 1), 0);

  IF (IsEmbrioDev = 1)
    THEN
    RETURN 7;
  END IF;


  
  IsTransfer = (SELECT
      1
    FROM v_ocells v
    WHERE v.IvfArtID = IvfArtID
    AND v.IvfTransferID IS NOT NULL
    LIMIT 1);

  
  IsFRZEmbrions = (SELECT
      1
    FROM v_ocells v
    WHERE v.IvfArtID = IvfArtID
    AND v.IvfCryoEggID IS NOT NULL
    LIMIT 1);

  
  IF (IsTransfer = 1
    AND IsFRZEmbrions = 1)
    THEN
    RETURN 12;
  END IF;

  
  IF (IsFRZEmbrions = 1)
    THEN
    RETURN 11;
  END IF;


  
  IF (IsTransfer = 1)
    THEN
    RETURN 10;
  END IF;

  
  IF (IsCancel = 1
    AND IvfFormID = 3)
    THEN
    RETURN 9;
  END IF;



  
  IsCancelAfterPunc :=
  COALESCE((SELECT
      1 AS b
    FROM ivf_stimulations s, ivf_punctions p
    WHERE s.IvfStimID = p.IvfStimID
    AND s.IvfArtID = IvfArtID
    AND NOT EXISTS (SELECT
        1
      FROM v_ocells o
      WHERE o.IvfArtID = s.IvfArtID
      AND o.IvfOStatusID != 4)), 0);

  IF (IsCancelAfterPunc = 1)
    THEN
    RETURN 8;
  END IF;



  
  IF (IsOldCanceled = 1)
    THEN
    IsCancel := CASE WHEN IsCancel = 0 THEN 1 ELSE  0 END ;
  END IF;


  
  
  IF (IsCancel = 1
    AND IvfFormID = 5)
    THEN
    RETURN 5;
  END IF;


  
  IF (IsCancel = 1
    AND IvfFormID = 3)
    THEN
    RETURN 9;
  END IF;

  IF (IsCancel = 1)
    THEN
    RETURN 6;
  END IF;

  IsEmbrCancel = (SELECT
      
      
      CASE WHEN COUNT(*) > 0      AND COUNT(*) = SUM(CASE WHEN iod.IvfOStatusID = 4 THEN 1 ELSE  0 END ) THEN 1 ELSE  0 END  AS Result

    FROM ivf_embryo_protocols iep
      INNER JOIN ivf_ocells_developments iod
        ON iep.EmbryoProtocolID = iod.EmbryoProtocolID

    WHERE iep.IvfArtID = IvfArtID);

  IF (IsEmbrCancel = 1) THEN
    RETURN 9;
  END IF;

  
  IsStim = COALESCE((SELECT
      1
    FROM ivf_stimulations s
    WHERE s.IvfArtID = IvfArtID), 0);


  
  IF (IsStim = 1
    AND IvfFormID = 5)
    THEN
    RETURN 4;
  END IF;

  
  IF (IsStim = 1)
    THEN
    RETURN 3;
  END IF;

  
  IF (IvfFormID = 2)
    THEN
    IsIschakvane = (SELECT
        1
      FROM v_ocells o
      WHERE o.IvfArtID = IvfArtID
      LIMIT 1);

    IF (IsIschakvane != 1)
      THEN
      RETURN 2;
    END IF;
  END IF;


  
  RETURN 1;

END;
$body$
LANGUAGE PLPGSQL
SECURITY DEFINER
;
-- REVOKE ALL ON FUNCTION f_artstatus (p_IvfArtID integer, IsCancel integer) FROM PUBLIC;
```

As I read, when result is get via SELECT, PostgreSQL don't permit SET of variable (see working example above).

If there is a only one result from SELECT:
```
name := (SELECT t.name from test_table t where t.id = x);
```
If there is a multiple results from SELECT:
```
select test_table.name into name from test_table where id = x;
```
I think that in ordinary example (when converting only @variable from MySQL), it is enough first form,
but the question is, is it possible with regex to check, that result will be assigned from SELECT and to remove "SET" keyword before variable when translate code with Ora2Pg?

**Comment by krasandr** [comment]
There is another one problem with conversion, this time with parameter list of MySQL function:

| f_converthormone | STRICT_TRANS_TABLES,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION | CREATE DEFINER=`root`@`%` FUNCTION `f_converthormone`(HormoneName VARCHAR(255), HormoneUnit VARCHAR(255), 
  HormoneValue double(17, 4)) RETURNS varchar(255) CHARSET utf8
BEGIN

  SET @ConvertOperation := 0;
  SET @ConvertFactor := 0;
  SELECT
    f.HormoneOperation,
    f.HormoneFactor INTO @ConvertOperation, @ConvertFactor
  FROM ivf_lab_formulas f
  WHERE f.HormoneName = HormoneName
  AND f.HormoneUnit = HormoneUnit;

  IF @ConvertOperation = 0 THEN 
    RETURN HormoneValue;
  END IF;

  IF @ConvertOperation = 1 THEN 
    RETURN HormoneValue + @ConvertFactor;
  END IF;

  IF @ConvertOperation = 2 THEN 
    RETURN HormoneValue - @ConvertFactor;
  END IF;

  IF @ConvertOperation = 3 THEN 
    RETURN HormoneValue * @ConvertFactor;
  END IF;

  IF @ConvertOperation = 4 THEN 
    RETURN HormoneValue / @ConvertFactor;
  END IF;

  RETURN 0;
END | utf8                 | utf8_general_ci      | latin1_swedish_ci  |

After conversion, PostgreSQL don't accept double precision (17,4) variable type:

CREATE OR REPLACE FUNCTION f_converthormone (HormoneName varchar(255), HormoneUnit varchar(255), HormoneValue double precision(17, 4)) RETURNS varchar AS $body$
DECLARE
ConvertOperation integer;
ConvertFactor integer;

BEGIN

  ConvertOperation := 0;
  ConvertFactor := 0;
  SELECT
    f.HormoneOperation,
    f.HormoneFactor INTO ConvertOperation, ConvertFactor
  FROM ivf_lab_formulas f
  WHERE f.HormoneName = HormoneName
  AND f.HormoneUnit = HormoneUnit;

  IF ConvertOperation = 0 THEN 
    RETURN HormoneValue;
  END IF;

  IF ConvertOperation = 1 THEN 
    RETURN HormoneValue + ConvertFactor;
  END IF;

  IF ConvertOperation = 2 THEN 
    RETURN HormoneValue - ConvertFactor;
  END IF;

  IF ConvertOperation = 3 THEN 
    RETURN HormoneValue * ConvertFactor;
  END IF;

  IF ConvertOperation = 4 THEN 
    RETURN HormoneValue / ConvertFactor;
  END IF;

  RETURN 0;
END;
$body$
LANGUAGE PLPGSQL
SECURITY DEFINER
;
-- REVOKE ALL ON FUNCTION f_converthormone (HormoneName varchar(255), HormoneUnit varchar(255), HormoneValue double precision(17, 4)) FROM PUBLIC;

**Comment by krasandr** [comment]
For previous function, I change double precision (17,4) with decimal (17,4) and it is created, but there is more complex problem with CAST of another function:

| f_getnextiznumber | STRICT_TRANS_TABLES,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION | CREATE DEFINER=`root`@`%` FUNCTION `f_getnextiznumber`(HospYear int, pIsAmbProc tinyint) RETURNS bigint(20)
BEGIN

  RETURN (SELECT
      IFNULL(MAX(CAST(LEFT(IZNumber, LENGTH(IZNumber) - 4) AS UNSIGNED)) + 1, 1) * 10000 + HospYear AS next_ize_with_year
    FROM department_patients dp
    WHERE RIGHT(IZNumber, 4) = HospYear
    AND dp.IsAmbProc = pIsAmbProc);

END | utf8                 | utf8_general_ci      | latin1_swedish_ci  |

Converted code with Ora2Pg simply miss "UNSIGNED" word and "AS" is empty, that is reason of getting mistake of CAST:

CREATE OR REPLACE FUNCTION f_getnextiznumber (HospYear integer, pIsAmbProc smallint) RETURNS bigint AS $body$
BEGIN

  RETURN(SELECT
      COALESCE(MAX(CAST(LEFT(IZNumber, LENGTH(IZNumber) - 4) AS)) + 1, 1) * 10000 + HospYear AS next_ize_with_year
    FROM department_patients dp
    WHERE RIGHT(IZNumber, 4) = HospYear
    AND dp.IsAmbProc = pIsAmbProc);

END;
$body$
LANGUAGE PLPGSQL
SECURITY DEFINER
;
-- REVOKE ALL ON FUNCTION f_getnextiznumber (HospYear integer, pIsAmbProc smallint) FROM PUBLIC;

**Comment by krasandr** [comment]
I finish converting of all MySQL functions (about 80, some of them are more complex) with Ora2Pg to PostgreSQL, and I just want to confirm to you, that mostly all problems was only one and the same, again and again - "SET" keyword before variable name, when assignment is from result of SELECT. Seems I miss it, when I write proposal of converting of @variables - when result is get from SELECT, PostgreSQL don't permit SET keyword for it.

**Comment by krasandr** [comment]
After I checked all functions, I think that enough fix will be to check with regex, is there a SELECT (case insensitive, with other words after it, followed by ";") and if there is, @variable from MySQL conversion will be converted to the same, but without @ in it's name, otherwise (if there is no SELECT), SET keyword will remain untouched before variable name (again without @ in it's name). It is very often problem in MySQL to get result for @variable via assignment, based on result (and type) of SELECT.
For me all situations in more than 80 functions was one and the same, no difference or exception of this rule.

**Comment by darold** [comment]
All these issue have been fixed with commits from ab779f0 to c6b3825, the only remaining issue is with multiple results from SELECT, this can not be detected for what I see.

**Comment by krasandr** [comment]
I confirm, that every issue of this topic is closed. Also, you are right - there is no way to know about multiple result set from SELECT, so it is not important. Thank you, and happy 30th birthday of Perl language today.
